In [1]:
import pandas as pd
import numpy as np
import plotly
import plotly.offline as py
import plotly.graph_objs as go
py.init_notebook_mode(connected=True)
import ipywidgets
from datetime import datetime
import bqplot as bq

In [2]:
%matplotlib inline

In [3]:
climate = pd.read_csv('https://query.data.world/s/w5y2cmvyy7yyckimsu4mtaduhpvxas')


In [4]:
climate.Latitude = climate.Latitude.map(lambda x: float(x[:-1]) if x[-1] == "N"  else  -float(x[:-1]))
climate.Longitude = climate.Longitude.map(lambda x: float(x[:-1]) if x[-1]== "E" else -float(x[:-1]))
climate.AverageTemperatureUncertainty = 9/5*climate.AverageTemperatureUncertainty+32
climate.AverageTemperature = 9/5*climate.AverageTemperature+32
climate.dt = climate.dt.map(lambda x: x[:-3])
climate.describe()

AverageTemperature  AverageTemperatureUncertainty       Latitude  \
count       228175.000000                  228175.000000  239177.000000   
mean            64.626744                      33.744818      22.532770   
std             18.044639                       1.763359      21.953405   
min            -16.189600                      32.072000     -37.780000   
25%             54.878000                      32.612000      12.050000   
50%             68.770400                      33.065600      28.130000   
75%             78.652400                      34.376000      37.780000   
max            100.909400                      57.266600      60.270000   

           Longitude  
count  239177.000000  
mean       43.720482  
std        64.951070  
min      -118.700000  
25%         4.050000  
50%        45.000000  
75%        99.910000  
max       151.780000

In [5]:
climate.shape

(239177, 7)

In [6]:
for column in climate.columns:
    try:
        missing_value = climate[column].isnull().value_counts()[1]
    except KeyError:
        print("In column {}, there is no missing value.".format(column))
    else:
        print("In column {}, there are {} missing value.".format(column, climate[column].isnull().value_counts()[1]))

In column dt, there is no missing value.
In column AverageTemperature, there are 11002 missing value.
In column AverageTemperatureUncertainty, there are 11002 missing value.
In column City, there is no missing value.
In column Country, there is no missing value.
In column Latitude, there is no missing value.
In column Longitude, there is no missing value.


### File Name: Global Land Temperature by Major City
***
### Data Source: [data.world - Global Climate Change Data](https://data.world/data-society/global-climate-change-data/workspace/file?filename=GlobalLandTemperatures%2FGlobalLandTemperaturesByMajorCity.csv)

***
### Data License: 
#### Subject to the terms and conditions of these Terms, and your compliance with these Terms, the Company hereby grants you a limited, personal, non-exclusive and non-transferable license to use and to display the Materials and to use the Site solely for your personal use. 
***
### Data Size: 13.48 MB (7 columns, 239,177 rows)
***
### Data Dictionary:
-  dt: data type, no missing data, the date of the temperature is recorded.
-  averagetemperature: decimal type, 11002 missing data, the average temperature of the location.
-  averagetemperatureuncertainty: decimal type, 11002 missing data the uncertainty of the average temperature of the location.
-  city: string type, no missing data, the name of the city.
-  country: string type, no missing data, the name of the country of the city in that row.
-  latitude: string type, no missing data, the latitude of the city.
-  longitude: string type, no missing data, the longitude of the city.



In [7]:
#handle missing values. Missing values constitute a very small percentage of the entire dataset
#replace values by 0
climate=climate.fillna(0)
climate

dt  AverageTemperature  AverageTemperatureUncertainty     City  \
0       1849-01             80.0672                        34.5830  Abidjan   
1       1849-02             81.3812                        34.4516  Abidjan   
2       1849-03             82.5818                        34.9016  Abidjan   
3       1849-04             79.0520                        34.4966  Abidjan   
4       1849-05             77.7686                        34.1600  Abidjan   
5       1849-06             76.7192                        34.5236  Abidjan   
6       1849-07             75.3044                        34.2572  Abidjan   
7       1849-08             74.4368                        34.2770  Abidjan   
8       1849-09             74.5916                        34.2068  Abidjan   
9       1849-10             77.4734                        34.1150  Abidjan   
10      1849-11             79.3976                        34.7126  Abidjan   
11      1849-12             77.8100                        35.3084  Abidjan   
12      1850-01             78.4454                        35.4974  Abidjan   
13      1850-02             82.2020                        34.5740  Abidjan   
14      1850-03             82.1336                        35.9114  Abidjan   
15      1850-04             79.7846                        34.9916  Abidjan   
16      1850-05             77.6822                        34.4390  Abidjan   
17      1850-06             76.8254                        34.1204  Abidjan   
18      1850-07             75.2720                        34.3418  Abidjan   
19      1850-08             74.7644                        34.3076  Abidjan   
20      1850-09             75.2000                        34.1960  Abidjan   
21      1850-10             76.0838                        34.1204  Abidjan   
22      1850-11             77.9756                        35.0942  Abidjan   
23      1850-12             78.8252                        34.8854  Abidjan   
24      1851-01             80.2202                        34.2482  Abidjan   
25      1851-02             81.2534                        34.1924  Abidjan   
26      1851-03             80.9258                        34.5704  Abidjan   
27      1851-04             79.8872                        34.9772  Abidjan   
28      1851-05             79.1816                        34.1510  Abidjan   
29      1851-06             76.8164                        34.3922  Abidjan   
...         ...                 ...                            ...      ...   
239147  2011-04             58.7768                        32.6534     Xian   
239148  2011-05             64.8050                        32.4788     Xian   
239149  2011-06             74.9822                        32.9594     Xian   
239150  2011-07             76.8110                        32.6174     Xian   
239151  2011-08             73.5242                        32.5418     Xian   
239152  2011-09             62.1950                        32.3744     Xian   
239153  2011-10             54.6566                        32.5292     Xian   
239154  2011-11             45.5774                        32.7038     Xian   
239155  2011-12             31.1180                        33.1952     Xian   
239156  2012-01             27.6512                        33.0224     Xian   
239157  2012-02             32.2916                        32.9090     Xian   
239158  2012-03             43.5740                        32.4482     Xian   
239159  2012-04             58.8344                        32.6876     Xian   
239160  2012-05             66.2738                        32.6696     Xian   
239161  2012-06             73.0058                        34.6064     Xian   
239162  2012-07             76.3628                        35.1500     Xian   
239163  2012-08             74.9930                        32.7830     Xian   
239164  2012-09             62.6504                        34.5290     Xian   
239165  2012-10             54.3182                        32.9

### use plotly and ipywidgets to plot geographical scatter graph by year

In [8]:
dates = pd.date_range(min(climate.dt), max(climate.dt), freq='M')

In [9]:
options = [(date.strftime('%Y-%m'), date) for date in dates]
index = (0, len(options)-1)

selection_range_slider = ipywidgets.widgets.SelectionSlider(
            options=options,
            continuous_update=False,
            value=options[-1][1],
    description='Date',
    orientation='horizontal',
    layout={'width': '800px'}
)


In [10]:
style = {'description_width': 'initial'}
longitude_slider=ipywidgets.widgets.SelectionSlider(
            options=range(-180,180),
            continuous_update = True,
            description='Map Center Longitude',
            orientation= 'horizontal',
            layout={'width': '400px'},
            style=style)

In [11]:
#*******************Plotting World Map*****************************

def plotmap(date="2013-08", middle_longitude=150):
    sample = climate[climate.dt == str(date)[0:7]]

    data = [ dict(
        lat = sample.Latitude.values,
        lon = sample.Longitude.values,
        text = sample.AverageTemperature.map(lambda x:str(x) + ' degree').values,
        marker = dict(
            color = sample.AverageTemperature.values,

           colorscale= "Hot",
            colorbar = dict(
                title = 'temperature',
                titleside = 'top',
                tickmode = 'array',
                tickvals = [30, 40,50,60,70,80,90,100,110,120],
                ticktext = ["30F", "40F", "50F", "60F", "70F", "80F", "90F", "100F", "110F", "120F"],
                ticks = 'outside'
            ),

            reversescale = True,
            opacity = 0.7,
            size = 10,

        ),
        type = 'scattergeo'
    ) ]

    layout = dict(
        geo = dict(
            scope='world',
            showland = True,
            landcolor = "rgb(212, 212, 212)",
            subunitcolor = "rgb(255, 255, 255)",
            countrycolor = "rgb(255, 255, 255)",
            showlakes = True,
            lakecolor = "rgb(255, 255, 255)",
            showsubunits = True,
            showcountries = True,
            resolution = 50,
            projection = dict(
                type = 'equirectangular',
                rotation = dict(
                    lon = middle_longitude
                )
            ),
            lonaxis = dict(
                showgrid = True,
                gridwidth = 0.5,
                range= [ -180.0, 180.0 ],
                dtick = 5
            ),
            lataxis = dict (
                showgrid = True,
                gridwidth = 0.5,
                range= [ -50.0, 70.0 ],
                dtick = 5
            )
        ),
        title = 'major city average temperature',
    )
    fig = { 'data':data, 'layout':layout }
    py.iplot(fig, filename='precipitation')
world_map= ipywidgets.interact(plotmap, middle_longitude = longitude_slider, date= selection_range_slider)
world_map

#*********************linegraph code************************

climate['dt'] = pd.to_datetime(climate['dt'])
climate['year'] = climate['dt'].dt.year
climate['month'] = climate['dt'].dt.month

is_India =  climate['Country']=='India'
climate_India = climate[is_India]

is_USA =  climate['Country']=='United States'
climate_USA = climate[is_USA]

is_China =  climate['Country']=='China'
climate_China = climate[is_China]

is_Canada =  climate['Country']=='Canada'
climate_Canada = climate[is_Canada]

is_Mexico =  climate['Country']=='Mexico'
climate_Mexico = climate[is_Mexico]

is_Australia =  climate['Country']=='Australia'
climate_Australia = climate[is_Australia]

is_Spain =  climate['Country']=='Spain'
climate_Spain = climate[is_Spain]

is_Russia =  climate['Country']=='Russia'
climate_Russia = climate[is_Russia]

is_Germany =  climate['Country']=='Germany'
climate_Germany = climate[is_Germany]

is_Italy =  climate['Country']=='Italy'
climate_Italy = climate[is_Italy]

d1= ipywidgets.Dropdown(
    options=["USA","India","China","Canada","Mexico","Australia","Spain","Russia","Germany","Italy"],
    value='USA',
    description='Country:',
)

def change_axis(d1):
    
    if d1 == "USA":
        x_data = climate_USA.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_USA.groupby(["year"])["AverageTemperature"].mean().values        
        x_sc = bq.LinearScale(min = 1770, max=2013)
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ['x','y'])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig1 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig1)


    if d1 == "India":
        x_data = climate_India.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_India.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig2 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig2)
        
    if d1 == "China":
        x_data = climate_China.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_China.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig3 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig3)    

   
    if d1 == "Canada":
        x_data = climate_Canada.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Canada.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig4 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig4)
        
    if d1 == "Mexico":
        x_data = climate_Mexico.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Mexico.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig5 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig5)
        
    if d1 == "Australia":
        x_data = climate_Australia.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Australia.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig6 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig6)
        
    if d1 == "Spain":
        x_data = climate_Spain.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Spain.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig7 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig7)
        
    if d1 == "Russia":
        x_data = climate_Russia.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Russia.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig8 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig8)
        
    if d1 == "Germany":
        x_data = climate_Germany.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Germany.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig9 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig9)
        
    if d1 == "Italy":
        x_data = climate_Italy.groupby(["year"])["AverageTemperature"].mean().index
        y_data = climate_Italy.groupby(["year"])["AverageTemperature"].mean().values
        x_sc = bq.LinearScale()
        y_sc = bq.LinearScale()
        ax_x = bq.Axis(label='Year', scale=x_sc)
        ax_y = bq.Axis(label='Average Temperature', scale=y_sc, orientation='vertical')
        tooltip = bq.Tooltip(fields = ["x", "y"])
        lines = bq.Lines(x=x_data, y=y_data, scales={'x': x_sc, 'y': y_sc},tooltip=tooltip, marker = 'circle')
        fig10 = bq.Figure(axes=[ax_x, ax_y], marks=[lines], title='Variation in Average Temperature between years 1743 - 2013: Country Level')
        display(fig10)
   
line_chart=ipywidgets.interact(change_axis, d1 = d1)
line_chart


interactive(children=(SelectionSlider(continuous_update=False, description='Date', index=3237, layout=Layout(w…

interactive(children=(Dropdown(description='Country:', options=('USA', 'India', 'China', 'Canada', 'Mexico', '…

<function __main__.change_axis(d1)>